django/contrib/auth/management/__init__.py

```diff
@@ -4,14 +4,104 @@
 from __future__ import unicode_literals
 
 import getpass
+import re
 import unicodedata
 
 from django.apps import apps as global_apps
 from django.contrib.auth import get_permission_codename
 from django.core import exceptions
-from django.db import DEFAULT_DB_ALIAS, router
+from django.db import DEFAULT_DB_ALIAS, migrations, router, transaction
 from django.utils import six
 from django.utils.encoding import DEFAULT_LOCALE_ENCODING
+from django.utils.text import camel_case_to_spaces
+
+
+class RenamePermissions(migrations.RunPython):
+    def __init__(self, app_label, old_model, new_model):
+        self.app_label = app_label
+        self.old_model = old_model
+        self.new_model = new_model
+        super(RenamePermissions, self).__init__(self.rename_forward, self.rename_backward)
+
+    def _rename(self, apps, schema_editor, backward=False):
+        Permission = apps.get_model('auth', 'Permission')
+        db = schema_editor.connection.alias
+        if not router.allow_migrate_model(db, Permission):
+            return
+
+        old_model_lower = self.old_model.lower()
+        new_model_lower = self.new_model.lower()
+
+        current_model = apps.get_model(app_label=self.app_label, model_name=new_model_lower)
+
+        # Build a dictionary allowing to do the correspondence between the current permission
+        # codenames and the new <codename, name> values.
+        correspondence = {}
+        for codename, name in _get_builtin_permissions(current_model._meta):
+            old_codename = re.sub(r'_%s$' % (new_model_lower), '_' + old_model_lower, codename)
+            if not backward:
+                correspondence[old_codename] = {'codename': codename, 'name': name}
+            else:
+                action = old_codename.split('_' + old_model_lower, 1)[0]
+                correspondence[codename] = {
+                    'codename': old_codename,
+                    'name': 'Can %s %s' % (action, camel_case_to_spaces(self.old_model))}
+
+        perms = Permission.objects.db_manager(db).filter(codename__in=correspondence).order_by('codename')
+        for perm in perms:
+            old_codename = perm.codename
+            perm.codename = correspondence[old_codename]['codename']
+            perm.name = correspondence[old_codename]['name']
+            with transaction.atomic(using=db):
+                perm.save(update_fields={'codename', 'name', })
+
+    def rename_forward(self, apps, schema_editor):
```

Comment:
I'd use rename_forwards/backwards for consistency with other methods like database_forwards.